enhancements/monitoring/alert-overrides.md

```diff
@@ -0,0 +1,560 @@
+- As an OpenShift user, I want to add labels to alerts to aide in routing and
+  triage.
+- As an OpenShift user, I want a supported way to add custom alerts using
+  metrics from components in the `openshift-*` namespaces.
+- As an OpenShift developer, I want the alerting rules shipped by my component
+  to be evaluated and forwarded to telemetry even when overridden by a user.
+
+### API Extensions
+
+We will be adding two new CRDs for alerting rules and relabeling:
+
+- `alertingrule.monitoring.openshift.io/v1alpha1`
+- `alertrelabelconfig.monitoring.openshift.io/v1alpha1`
+
+```go
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+
+// AlertingRule represents a set of user-defined Prometheus rule groups containing
+// alerting rules -- recording rules are not allowed.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +k8s:openapi-gen=true
+type AlertingRule struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the desired state of this AlertingRule object.
+	Spec AlertingRuleSpec `json:"spec"`
+
+	// status describes the current state of this AlertOverrides object.
+	//
+	// +optional
+	Status AlertingRuleStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AlertingRuleList is a list of AlertingRule objects.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +k8s:openapi-gen=true
+type AlertingRuleList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// items is a list of AlertingRule objects.
+	Items []AlertingRule `json:"items"`
+}
+
+// AlertingRuleSpec is the desired state of an AlertingRule resource.
+//
+// +k8s:openapi-gen=true
+type AlertingRuleSpec struct {
+	// groups is a list of grouped alerting rules.
+	Groups []RuleGroup `json:"groups,omitempty"`
+}
+
+// RuleGroup is a list of sequentially evaluated alerting rules.
+//
+// +k8s:openapi-gen=true
+type RuleGroup struct {
+	// name is the name of the group.
+	Name string `json:"name"`
+
+	// interval is how often rules in the group are evaluated.  If not specified,
+	// it defaults to the global.evaluation_interval configured in Prometheus,
+	// which itself defaults to 1 minute.
+	//
+	// +optional
+	Interval string `json:"interval,omitempty"`
+
+	// rules is a list of sequentially evaluated alerting rules.
+	Rules []Rule `json:"rules"`
```

Comment:
Does this need to have at least 1 entry? If so you might want
```
// +kubebuilder:validation:MinLength:=1
```
What happens if this list is empty? Is it possible if this is empty to create a valid Alerting rule on prometheus?

Reply:
It seems weird to me that you'd create a rule group with no rules, but if there's a genuine customer use case here I guess we can allow it.

You'll need an `// +optional` tag and `omitempty`

I guess the question is, do we expect a user to know what to add here? The main reason we add these validations is to help a user to understand they've configured it correctly, so, are we going to have users thinking adding a name to this list is enough to do something?

Reply:
I just made this required with a min length of one. It shouldn't be empty really.